flink-filesystems/flink-s3-fs-presto/src/main/java/org/apache/flink/fs/s3presto/FlinkS3PrestoFileSystem.java

```diff
@@ -0,0 +1,164 @@
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.fs.s3.common.FlinkS3FileSystem;
+import org.apache.flink.fs.s3.common.writer.S3AccessHelper;
+import org.apache.flink.util.ExceptionUtils;
+
+import org.apache.hadoop.fs.FileSystem;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.nio.file.DirectoryNotEmptyException;
+import java.util.Optional;
+
+/**
+ * {@code FlinkS3PrestoFileSystem} provides custom recursive deletion functionality to work around a
+ * bug in the internally used Presto file system.
+ *
+ * <p>https://github.com/prestodb/presto/issues/17416
+ */
+public class FlinkS3PrestoFileSystem extends FlinkS3FileSystem {
+
+    private static final Logger LOG = LoggerFactory.getLogger(FlinkS3PrestoFileSystem.class);
+
+    /**
+     * Creates a {@code FlinkS3PrestoFileSystem} based on the given Hadoop S3 file system. The given
+     * Hadoop file system object is expected to be initialized already.
+     *
+     * <p>This constructor additionally configures the entropy injection for the file system.
+     *
+     * @param hadoopS3FileSystem The Hadoop FileSystem that will be used under the hood.
+     * @param entropyInjectionKey The substring that will be replaced by entropy or removed.
+     * @param entropyLength The number of random alphanumeric characters to inject as entropy.
+     */
+    public FlinkS3PrestoFileSystem(
+            FileSystem hadoopS3FileSystem,
+            String localTmpDirectory,
+            @Nullable String entropyInjectionKey,
+            int entropyLength,
+            @Nullable S3AccessHelper s3UploadHelper,
+            long s3uploadPartSize,
+            int maxConcurrentUploadsPerStream) {
+        super(
+                hadoopS3FileSystem,
+                localTmpDirectory,
+                entropyInjectionKey,
+                entropyLength,
+                s3UploadHelper,
+                s3uploadPartSize,
+                maxConcurrentUploadsPerStream);
+    }
+
+    @Override
+    public boolean delete(Path path, boolean recursive) throws IOException {
+        if (recursive) {
+            deleteRecursively(path);
+        } else {
+            if (isDirectoryWithContent(path)) {
+                throw new DirectoryNotEmptyException(path.getPath());
+            }
+
+            deleteObject(path);
+        }
+
+        return true;
+    }
+
+    private void deleteRecursively(Path path) throws IOException {
+        final FileStatus[] containingFiles = listStatus(path);
+        if (containingFiles == null) {
+            LOG.warn(
+                    "No files could be retrieved even though the path was marked as existing. "
+                            + "This is an indication for a bug in the underlying FileSystem "
+                            + "implementation and should be reported. It won't affect the "
+                            + "processing of this run, though.");
```

Comment:
If it is not a problem, why log it?

This message is also not comprehensible for users; it doesn't mention listing files or even directories.

Reply:
It also doesn't make sense to just return in this case; we have no idea what the actual external state is because we already assume the underlying FS to be broken.

Reply:
fair enough: The contract of the HadoopFileSystem.listStatus states that it should never return `null` (see [JavaDoc](https://hadoop.apache.org/docs/stable/api/org/apache/hadoop/fs/FileSystem.html#listStatus-org.apache.hadoop.fs.Path-)). Therefore, doing a null check as a precondition should be good enough. 👍